Support newer versions of LLVM and ROCm

tbennun opened the pull request:
Adds support for newer versions of LLVM (18) and ROCm (6.2).

tbennun commented on lib/JitEngine.cpp:
```diff
@@ -42,7 +43,11 @@ using namespace llvm;
 JitEngine::createTargetMachine(Module &M, StringRef Arch, unsigned OptLevel) {
   Triple TT(M.getTargetTriple());
+#if LLVM_VERSION_MAJOR >= 18
+  std::optional<llvm::CodeGenOptLevel> CGOptLevel = CodeGenOpt::getLevel(OptLevel);
```
One could also just use `auto` here

ggeorgakoudis replied:
Let's go for `auto` to avoid preprocessor pollution

tbennun replied:
Done.

tbennun commented:
> We need to have tests ran for LLVM 18 too. However, github and tioga tests succeeded on LLVM 17, which makes me believe that your changes are portable for both LLVM 17 and 18?

The functions I'm using (`PointerType::get`) are compatible going back to LLVM 14 (and perhaps even earlier).

I also addressed your comment and it still works for me.

ggeorgakoudis commented:
Supereded #44
